torch/distributed/_composable/fsdp/_fsdp_param.py

```diff
@@ -609,13 +595,24 @@ def alloc_all_gather_outputs(self) -> None:
             alloc_storage(tensor)
 
     def free_unsharded_param(self) -> None:
-        for tensor in itertools.chain(
-            self.all_gather_outputs, self._unsharded_inner_tensors
-        ):
-            free_storage(tensor)
         if ca.compiled_autograd_enabled:
+            """
+            Assumptions under compile:
+            - `self._unsharded_param` is NOT an alias of `self.all_gather_outputs`.
+            Instead, we explicitly *copy* the data from `self.all_gather_outputs`
+            to `self._unsharded_param` in `init_unsharded_param()`.
+            - `self.all_gather_outputs` and `self._unsharded_inner_tensors` are NOT
+            graph inputs. They are created within the graph and is guaranteed to be freed
+            by the end of the graph. They don't leak outside of the graph.
+            """
+            free_storage(self._unsharded_param)
             self.all_gather_outputs = []
             self._unsharded_inner_tensors = []
+        else:
+            for tensor in itertools.chain(
+                self.all_gather_outputs, self._unsharded_inner_tensors
+            ):
+                free_storage(tensor)
```

Comment:
hmm this reminds me - do the mutation barrier mechanisms that we have in our graph passes today respect the `inductor::storage_resize_bytes_` op that this eventually calls?

At first I was going to say no, because that op secretly advertises as having a functional schema ([code](https://github.com/pytorch/pytorch/blob/main/torch/csrc/inductor/resize_storage_bytes.cpp#L55))

But on a second look, maybe yes because.. the mutation barrier logic looks like it just checks for a trailing `_` in the function name to tell if the op is mutable? ([code](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/pattern_matcher.py#L1605))

Reply:
yeah I think it just checks for trailing `_` right now.. I poked at that a few weeks ago and I think that infra needs to work with non-OpOverload ops too, that's why it's using a complicated regex :(